File: Practice/AllDomains/CoreCS/Algorithms/Strings/GameOfThronesII.js
/*	author:@shivkrthakur  */
function modularMultiplication(a, b, mod) {
    var radix = 31623; 
    var ra = a % radix;
    var rb = b % radix;
    var da = Math.floor(a / radix);
    var db = Math.floor(b / radix);
    var res = 0;
    var tmp = (((da * db) % mod) * ((radix * radix) % mod)) % mod;
    res += tmp;
    tmp = (((((da * rb) % mod) + ((db * ra) % mod)) % mod) * radix) % mod;
    res += tmp;
    res %= mod;
    tmp = (ra * rb) % mod;
    res += tmp;
    res %= mod;
    return res;
}

function modularDivision(a, b, m){
    
    return modularMultiplication(a,modularExponentiation(b, (m - 2), m), m);
}

function modularExponentiation(x, y, m){
    var result = 1;
    x = x % m;
    while(y > 0) {
        if((y & 1) > 0) result = modularMultiplication(result,x, m);
        y >>= 1;
        x = modularMultiplication(x,x, m);
    }
    return result;
}

function recursiveFactorial(val, mod){
    if(val == 0) return 1;
    var temp = factorial(val -1, mod) % mod;
    return ((val % mod) * (temp % mod)) % mod;
}

function factorial(val, mod){
    var result = 1;
    for(var i = val; i > 0; i--){
        result = modularMultiplication(result, i, mod);
    }
    return result;
}

function processData(str) {
    //Enter your code here
    var mod = 1000000007;
    var char = [];
    for(var k = 0; k < str.length; k++){
      var index = str.charCodeAt(k) - 'a'.charCodeAt(0);
      if(char[index]) char[index]++;
      else char[index] = 1;
    }
    var den = 1;
    for(var l = 0; l < char.length; l++){
      if(char[l]) 
      {
          var temp = parseInt(char[l]/2);
          var fact = factorial(temp, mod) % mod;
          den = modularMultiplication(den, fact, mod);
      }
    }
    var sum = parseInt(str.length/2);
    var sumFact = factorial(sum, mod);
    //console.log('sum: ' + sum + ' sumFact: ' + sumFact + ' den: ' + den);
    console.log((modularDivision(sumFact,den, mod)));
} 

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input);
});

File: Practice/AllDomains/CoreCS/Algorithms/Strings/GameOfThronesII.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
using System.Numerics;
class Solution 
{
    static void Main(String[] args) 
	{
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution */
        string s = Console.ReadLine();
        int mod7 = (int)(Math.Pow(10,9) + 7);
        int [] chars = new int[26];
        for(int i = 0; i < s.Length; i++) chars[s[i] - 'a']++;
        long num = s.Length/2, den = 1;
        for(int i = 0; i < chars.Length; i++)
        {
            if(chars[i] > 0)
            {
                int temp = chars[i]/2;
                long fact = Factorial(temp, mod7) % mod7;
                den = ((den % mod7) * (fact % mod7)) % mod7;
            }
        }
        long numFact = Factorial(num, mod7) % mod7;
        long output = DivisionModPrimeUsingFermentsLT(numFact, den, mod7) % mod7;
        Console.WriteLine(output);
    }
    
    static long Factorial(long num, int mod7)
    {
        if(num == 0) return 1;
        long temp = Factorial(num - 1, mod7) % mod7;
        return ((num % mod7) * (temp % mod7)) % mod7;
    }
    
    static long DivisionModPrimeUsingFermentsLT(long num, long den, int mod)
    {
        long answer = num * (ModularExponentiations(den, (mod - 2), mod));
        return answer;
    }

    static long ModularExponentiations(long x, long y, int p)
    {
        long result = 1;
        x = x % p;

        while(y > 0)
        {
            if((y & 1) > 0) result = (result * x) % p;
            y >>= 1;
            x = (x * x) % p;
        }
        return result;
    }
}